sdk/include/opentelemetry/sdk/trace/sampler.h

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include "opentelemetry/version.h"
+#include "opentelemetry/common/attribute_value.h"
+
+#include <memory>
+#include <map>
+#include <string>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+class Sampler
+{
+public:
+  /**
+   * Initialize a new tracer.
+   * @param processor The span processor for this tracer. This must not be a
+   * nullptr.
+   */
+  virtual Decision shouldSample();
```

Comment:
The [list of arguments for this function](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#shouldsample) will be crucial. I will add some links here, so you can infer the proper types.

> * SpanContext of a parent Span. Typically extracted from the wire. Can be null.

This is a work in progress in [another PR](https://github.com/open-telemetry/opentelemetry-cpp/pull/42/files#diff-4481fa0d188c437f57f7ed10955f28abR34). For now it might be best to add an empty `SpanContext` class and use it as a placeholder. Probably best wrapped in a `shared_ptr`, so it can be nullable.


> * TraceId of the Span to be created. It can be different from the TraceId in the SpanContext. Typically in situations when the Span to be created starts a new Trace.

This will be of the [`TraceId`](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/trace/trace_id.h#L30) type.


> * Name of the Span to be created.

This will be passed on as a `string_view`, as it goes in [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/trace/tracer.h#L26).

> * SpanKind

This will come from [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/trace/span.h#L44).

> * Initial set of Attributes for the Span being constructed

We should keep this in sync with [this attribute](https://github.com/open-telemetry/opentelemetry-cpp/pull/117/files#diff-ee947b76b7c7a798b53dd096a45166a9R44), which will most likely passed on as is.

> * Collection of links that will be associated with the Span to be created. Typically useful for batch operations, see Links Between Spans.

We don't have links yet, so let's leave this out for now.